### ISSUE TITLE
Model export fails when excluding fields from a List of nested models

### ISSUE THREAD

**Issue by masalim2** [samuelcolvin/pydantic#1283]
# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.4
            pydantic compiled: False
                 install path: /Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic
               python version: 3.7.4 (default, Sep  7 2019, 18:27:02)  [Clang 10.0.1 (clang-1001.0.46.4)]
                     platform: Darwin-18.7.0-x86_64-i386-64bit
     optional deps. installed: []
```

Hi all, thanks for the terrific library -- it has been a pleasure to use so far.  I am attempting to export a nested model via ``dict()`` with the ``exclude`` kwarg to omit some fields in the inner child model.  This ``exclude`` seems to cause a ``TypeError`` when the exclusion field pertains to a *list* of nested models.  

I wrote a simple reproducer for the issue below.  I am encountering a TypeError inside the the ``_normalize_indexes`` method of ``pydantic.utils.ValueItems``. 

```py
from pydantic import BaseModel
from typing import List

class Foo(BaseModel):
    id: int

class Bar(BaseModel):
    id: int
    foos: List[Foo] = []

my_bar = Bar(id=3)
my_bar.dict()  # This is ok 
my_bar.dict(exclude={"foos": {"id"}}) # This causes TypeError
```

The Traceback is:
```
Traceback (most recent call last):
  File "tester.py", line 13, in <module>
    my_bar.dict(exclude={"foos": {"id"}})
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/main.py", line 343, in dict
    exclude_none=exclude_none,
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/main.py", line 333, in <dictcomp>
    return {
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/main.py", line 675, in _iter
    exclude_none=exclude_none,
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/main.py", line 587, in _get_value
    value_exclude = ValueItems(v, exclude) if exclude else None
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/utils.py", line 295, in __init__
    items = self._normalize_indexes(items, len(value))
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/utils.py", line 345, in _normalize_indexes
    return {v_length + i if i < 0 else i for i in items}
  File "/Users/misha/workflow/balsam/env/lib/python3.7/site-packages/pydantic/utils.py", line 345, in <setcomp>
    return {v_length + i if i < 0 else i for i in items}
TypeError: '<' not supported between instances of 'str' and 'int'
```

Dropping in with a debugger, it looks like a ``ValueItems`` is constructed twice.  The first time around is fine, where the local variables in ``ValueItems.__init__`` are ``items = {'foos': {'id'}}`` and ``value = Bar(id=3, foos=[])``.  

The second time around, ``items = {'id'}`` and ``value = []`` causes a problem.   It seems like the issue is here in the ``ValueItems.__init__`` logic:  https://github.com/samuelcolvin/pydantic/blob/e3243d267b06bda2d7b2213a8bad70f82171033c/pydantic/utils.py#L347-L348  

Inside the set comprehension of ``self._normalize_indexes``, the comparison `i < 0` is taking place between the string ``"id"`` and literal int 0.  This is not an edge case for the empty list; I find the same traceback occurs when I repeat the example with a non-empty list of ``Foo`` objects.

Should it be that ``value``, not ``items`` is passed as the first argument to ``self._normalize_indexes`` here?  When I make that change locally, the test seems to pass but I have no familiarity with the codebase outside of this little exploration.

**Comment by masalim2** [comment]
I just found a relevant example after searching a little deeper in the repository: 

https://github.com/samuelcolvin/pydantic/blob/e3243d267b06bda2d7b2213a8bad70f82171033c/docs/examples/exporting_models_exclude2.py#L54-L55

It looks like the right way to exclude fields from nested List elements is to provide an exclusion set **per-index**.

To update my failing code snippet:
```python3
from pydantic import BaseModel
from typing import List

class Foo(BaseModel):
    id: int
    name: str = "foo-name"

class Bar(BaseModel):
    id: int
    foos: List[Foo] = []

my_bar = Bar(id=3, foos=[Foo(id=1), Foo(id=2), Foo(id=3)])
# my_bar.dict(exclude={"foos": {"id"}})  # This causes TypeError

# The right way to do it:
foo_excludes = {idx: {"id"} for idx in range(len(my_bar.foos))}
my_bar.dict(exclude={"foos": foo_excludes})
```

I would suggest raising an exception with more informative message when the wrong types end up being passed into ``ValueItems`` on a model export call.  The behavior of ``include`` and ``exclude`` for list and tuple fields should also be documented in the docs section https://pydantic-docs.helpmanual.io/usage/exporting_models/#advanced-include-and-exclude

That being said, having to construct key-value pairs in the exclusion dictionary element-wise feels a bit heavy-handed.  Perhaps we can consider:

1.  Supporting non-indexed inclusion/exclusion of fields in a List or Tuple: when a dict with ``str`` keys is passed for one of the inclusion/exclusion fields, apply the same logic to each element of the sequence.
2.  Providing **default** include or exclude fields in a model's ``Config`` class, so that nested submodels can define their export behavior without having to build up deeply nested dictionary kwargs

**Comment by AlexECX** [comment]
Just encountered this bug too. My model has multiple lists of nested models that I need to exclude read-only fields from, so I was planning to build a ReadOnly filter once and re-use it on every call. Not needing to dynamically create the filter for every instance would be practical.

**Comment by samuelcolvin** [comment]
I'd be happy to accept a PR to:
* provide a more helpful error
* document the behaviour
* provide an easy way to exclude a given element from all members of a list/tuple, .e.g `__all__`.

Excluding field by default based on config is a possibility, but not a no-brainer; might be related to (or could be implemented as part of) #624.

**Comment by masalim2** [comment]
Okay, I've added the ability to use ``__all__`` to exclude a given field from all members of a list/tuple and updated PR #1286 accordingly.

**Comment by masalim2** [comment]
Closed in #1286
